sqldb/migrations_test.go

```diff
@@ -452,3 +452,121 @@ func TestCustomMigration(t *testing.T) {
 		})
 	}
 }
+
+// TestSchemaMigrationIdempotency tests that the our schema migrations are
+// idempotent. This means that we can apply the migrations multiple times and
+// the schema version will always be the same.
+func TestSchemaMigrationIdempotency(t *testing.T) {
+	dropMigrationTrackerEntries := func(t *testing.T, db *BaseDB) {
+		_, err := db.Exec("DELETE FROM migration_tracker;")
+		require.NoError(t, err)
+	}
+
+	lastMigration := migrationConfig[len(migrationConfig)-1]
+
+	t.Run("SQLite", func(t *testing.T) {
+		// First instantiate the database and run the migrations
+		// including the custom migrations.
+		t.Logf("Creating new SQLite DB for testing migrations")
+
+		dbFileName := filepath.Join(t.TempDir(), "tmp.db")
+		var (
+			db  *SqliteStore
+			err error
+		)
+
+		// Run the migration 3 times to test that the migrations
+		// are idempotent.
+		for i := 0; i < 3; i++ {
+			db, err = NewSqliteStore(&SqliteConfig{
+				SkipMigrations: false,
+			}, dbFileName)
+			require.NoError(t, err)
+
+			dbToCleanup := db.DB
+			t.Cleanup(func() {
+				require.NoError(
+					t, dbToCleanup.Close(),
+				)
+			})
+
+			ctxb := context.Background()
+			require.NoError(
+				t, db.ApplyAllMigrations(ctxb, GetMigrations()),
+			)
+
+			version, dirty, err := db.GetSchemaVersion()
+			require.NoError(t, err)
+
+			// Now reset the schema version to 0 and make sure that
+			// we can apply the migrations again.
+			require.Equal(t, lastMigration.SchemaVersion, version)
+			require.False(t, dirty)
+
+			db.SetSchemaVersion(-1, false)
```

Comment:
Q: why -1 and not 0 ?

Reply:
This is what golang migrate will return by default if there's no tracked schema version. Replaced it with `database.NilVersion` for better readability.